x-pack/metricbeat/module/iis/application_pool/reader.go

```diff
@@ -62,87 +63,138 @@ var appPoolCounters = map[string]string{
 }
 
 // newReader creates a new instance of Reader.
-func newReader() (*Reader, error) {
+func newReader(config Config) (*Reader, error) {
 	var query pdh.Query
 	if err := query.Open(); err != nil {
 		return nil, err
 	}
-	reader := &Reader{
-		Query: query,
-		log:   logp.NewLogger("website"),
+	r := &Reader{
+		query:           query,
+		log:             logp.NewLogger("application_pool"),
+		config:          config,
+		workerProcesses: make(map[string]string),
+	}
+	apps, err := getApplicationPools(config.Names)
+	if err != nil {
+		return r, errors.Wrap(err, "failed retrieving running worker processes")
+	}
+	if len(apps) == 0 {
+		r.log.Info("no running application pools found")
+		return r, nil
+	}
+	r.applicationPools = apps
+
+	for key, value := range appPoolCounters {
+		childQueries, err := query.GetCounterPaths(value)
+		if err != nil {
+			if err == pdh.PDH_CSTATUS_NO_COUNTER || err == pdh.PDH_CSTATUS_NO_COUNTERNAME || err == pdh.PDH_CSTATUS_NO_INSTANCE || err == pdh.PDH_CSTATUS_NO_OBJECT {
+				r.log.Infow("Ignoring non existent counter", "error", err,
+					logp.Namespace("application pool"), "query", value)
+				continue
+			} else {
+				return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, value)
+			}
+		}
+		// check if the pdhexpandcounterpath/pdhexpandwildcardpath functions have expanded the counter successfully.
+		if len(childQueries) == 0 || (len(childQueries) == 1 && strings.Contains(childQueries[0], "*")) {
+			// covering cases when PdhExpandWildCardPathW returns no counter paths or is unable to expand and the ignore_non_existent_counters flag is set
+			r.log.Debugw("No counter paths returned but PdhExpandWildCardPathW returned no errors", "initial query", value,
+				logp.Namespace("application pool"), "expanded query", childQueries)
+			continue
+		}
+		for _, v := range childQueries {
+			if err := query.AddCounter(v, "", "float", len(childQueries) > 1); err != nil {
+				return nil, errors.Wrapf(err, `failed to add counter (query="%v")`, v)
+			}
+			r.workerProcesses[v] = key
+		}
```

Comment:
Code added here is mostly the same as in `refreshCounterPaths()`, could it be reused?

Reply:
refactored the code so the function is reused